File: backend/shortest-path/precompute.py
import re
import json
from collections import defaultdict
from geopy.distance import geodesic
from pymongo import MongoClient
import os 

coord_pattern = re.compile(r"\((.+)\)")

client = MongoClient('mongodb+srv://alhassan:p7OcHSJ3LRpJTgHz@stations.r5pneer.mongodb.net/stations?retryWrites=true&w=majority')
db = client.stations
stops_collection = db.stations
routes_collection = db.routes

stops = {}
for stop in stops_collection.find():
    coordinate_string = coord_pattern.search(stop['geometry']).group(1).strip()
    coordinates = tuple(float(coord) for coord in coordinate_string.split(' ') if coord)

    stops[stop['stop_name']] = {'coordinates': coordinates, 'connections': defaultdict(list)}

def calculate_distance(coord1, coord2):
    return geodesic(coord1, coord2).kilometers

for row in routes_collection.find():
    route_coordinates = [tuple(map(float, coords.split())) for coords in coord_pattern.findall(row['geometry'])[0].split(",")]
    for i in range(len(route_coordinates) - 1):
        start_coords, end_coords = route_coordinates[i], route_coordinates[i+1]
        start_station = min(stops, key=lambda s: calculate_distance(stops[s]['coordinates'], start_coords))
        end_station = min(stops, key=lambda s: calculate_distance(stops[s]['coordinates'], end_coords))

        if start_station != end_station:
            if end_station not in stops[start_station]['connections']:
                stops[start_station]['connections'][end_station] = {'route_id': row['route_id'], 'distance': calculate_distance(stops[start_station]['coordinates'], stops[end_station]['coordinates'])}
            if start_station not in stops[end_station]['connections']:
                stops[end_station]['connections'][start_station] = {'route_id': row['route_id'], 'distance': calculate_distance(stops[start_station]['coordinates'], stops[end_station]['coordinates'])}

if os.path.exists('metro_graph.json'):
    os.remove('metro_graph.json')

with open('metro_graph.json', 'w') as f:
    json.dump(stops, f)

File: backend/Metro-Shortest-Path/precompute.js
const mongoose = require('mongoose');
const geolib = require('geolib');
const fs = require('fs');
const path = require('path');

const jsonDirectory = path.join(process.cwd(), 'json');

module.exports = function precompute() {

// Regular expression pattern for coordinates
const coordPattern = /\((.+)\)/;

// Connect to MongoDB
mongoose.connect('mongodb+srv://alhassan:p7OcHSJ3LRpJTgHz@stations.r5pneer.mongodb.net/stations?retryWrites=true&w=majority', { useNewUrlParser: true, useUnifiedTopology: true });

const db = mongoose.connection;
const stopsCollection = db.collection('stations');
const routesCollection = db.collection('routes');

let stops = {};

stopsCollection.find().toArray((err, stopResults) => {
    stopResults.forEach(stop => {
        let coordinateString = coordPattern.exec(stop.geometry)[1].trim();
        let coordinates = coordinateString.split(' ').map(coord => parseFloat(coord)).filter(coord => coord);

        stops[stop.stop_name] = { 'coordinates': coordinates, 'connections': {} };
    });

    function calculateDistance(coord1, coord2) {
        return geolib.getDistance(
            { latitude: coord1[1], longitude: coord1[0] },
            { latitude: coord2[1], longitude: coord2[0] },
            1
        ) / 1000; // Converting from meters to kilometers
    }

    routesCollection.find().toArray((err, routeResults) => {
        routeResults.forEach(row => {
            let routeCoordinates = coordPattern.exec(row.geometry)[1].split(",").map(coords => coords.split().map(coord => parseFloat(coord)));

            for (let i = 0; i < routeCoordinates.length - 1; i++) {
                let startCoords = routeCoordinates[i], endCoords = routeCoordinates[i + 1];
                let startStation = Object.keys(stops).reduce((a, b) => calculateDistance(stops[a].coordinates, startCoords) < calculateDistance(stops[b].coordinates, startCoords) ? a : b);
                let endStation = Object.keys(stops).reduce((a, b) => calculateDistance(stops[a].coordinates, endCoords) < calculateDistance(stops[b].coordinates, endCoords) ? a : b);

                if (startStation !== endStation) {
                    if (!stops[startStation].connections[endStation]) {
                        stops[startStation].connections[endStation] = { 'route_id': row.route_id, 'distance': calculateDistance(stops[startStation].coordinates, stops[endStation].coordinates) };
                    }
                    if (!stops[endStation].connections[startStation]) {
                        stops[endStation].connections[startStation] = { 'route_id': row.route_id, 'distance': calculateDistance(stops[startStation].coordinates, stops[endStation].coordinates) };
                    }
                }
            }
        });

        // Save the output to a JSON file
        fs.writeFileSync(jsonDirectory + '/metro_graph.json', JSON.stringify(stops, null, 4));
    });
});
}